kernel/cClosure.ml

```diff
@@ -335,19 +337,22 @@ type clos_infos = {
   i_flags : reds;
   i_cache : infos_cache }
 
-type clos_tab = fconstr KeyTable.t
+type clos_tab = fconstr constant_def KeyTable.t
 
 let info_flags info = info.i_flags
 let info_env info = info.i_cache.i_env
 
 (**********************************************************************)
 (* The type of (machine) stacks (= lambda-bar-calculus' contexts)     *)
+type 'a next_native_args = (CPrimitives.arg_kind * 'a) list
 
 type stack_member =
   | Zapp of fconstr array
   | ZcaseT of case_info * constr * constr array * fconstr subs
   | Zproj of Projection.Repr.t
   | Zfix of fconstr * stack
+  | Zprimitive of CPrimitives.t * pconstant * fconstr list * fconstr next_native_args
```

Comment:
IIUC primitives are never polymorphic so this should be a `Constant.t` instead of pconstant.

Reply:
Also, why are you using list when the others are using array?

Reply:
I'm not so sure about `Constant.t`. If I use it, then each time I do `mkConstant`, I'm relying on the fact that the constant is not polymorphic. If I use `pconstant`, then I encode this invariant in the type at construction time: it is a `pconstant`, with an empty instance.

In fact, I thought we were not supposed to use `mkConstant` anymore. Isn't it dangerous to insert empty instances in so many places?

Reply:
> Also, why are you using list when the others are using array?

What do you mean by "the others"? I believe the code needs to split this list at places where arguments need to be reduced before the primitive can be evaluated. So lists seem like a natural choice, don't they?

Reply:
>In fact, I thought we were not supposed to use mkConstant anymore. Isn't it dangerous to insert empty instances in so many places?

Fair enough.